### PR TITLE
restrict wazo-dird /config endpoints to master tenant

### PR DESCRIPTION
**Pull request by nissou91**
* We want to restrict the `config` endpoint in `wazo-dird` to be accessed in `read-only` mode only by the `master_tenant` , an `HTTP 401` error is raised otherwise.

Depends-On: https://github.com/wazo-platform/wazo-dird-client/pull/17

### CODE REVIEW

**Comment by nissou91** [comment]
recheck

**Comment by sduthil** [comment]
recheck

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,15 @@
...
 
+## 22.08
+
+* The `config` endpoint is now accessible in read-only mode by:
+
+  * `master_tenant`
+
+* The `status` endpoint has now a new route:
+
+  * `/config`
```
Comment (fblackburn1) [inline CHANGELOG.md:11]: What does it mean?

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,15 @@
 # Changelog
 
+## 22.08
+
+* The `config` endpoint is now accessible in read-only mode by:
+
+  * `master_tenant`
```
Comment (fblackburn1) [inline CHANGELOG.md:7]:
```suggestion
* The `config` endpoint is now accessible only by *master* tenant
```

File: wazo_dird/controller.py [hunk 3]
```diff
@@ -30,6 +30,16 @@ def __init__(self, config):
...
         self.token_renewer = TokenRenewer(self.auth_client)
+
+        if not self.config['auth'].get('master_tenant_uuid'):
+            self.token_renewer.subscribe_to_next_token_details_change(
+                auth.init_master_tenant
+            )
+
+        self.token_renewer.subscribe_to_next_token_details_change(
+            lambda t: self.token_renewer.emit_stop()
+        )
```
Comment (fblackburn1) [inline wazo_dird/controller.py:41]: Why did you stop the token renewer?

File: wazo_dird/plugins/config/http.py [hunk 4]
```diff
@@ -1,18 +1,21 @@
...
+from wazo_dird.http import AuthResource
+from xivo.auth_verifier import required_acl
 
 
-class Config(LegacyAuthResource):
+class Config(AuthResource):
 
     _config_service = None
 
+    @required_master_tenant()
```
Comment (fblackburn1) [inline wazo_dird/plugins/config/http.py:13]: Why?

File: wazo_dird/plugins/status/http.py [hunk 5]
```diff
@@ -6,9 +6,23 @@
...
+            'rest_api': {
+                'status': self.status_aggregator.status()['rest_api']['status']
+            },
+            'master_tenant': {
+                'status': 'ok'
+                if self._config['auth'].get('master_tenant_uuid')
+                else 'fail'
+            },
+        }
+        return result, 200
```
Comment (fblackburn1) [inline wazo_dird/plugins/status/http.py:28]: Why? Please read how status_aggregator work
You can look at wazo-chatd and/or almost all other services

File: wazo_dird/plugins/status/plugin.py [hunk 6]
```diff
@@ -14,12 +14,14 @@ class StatusViewPlugin(BaseViewPlugin):
...
 
+        config = dependencies['config']
+
         status_aggregator = dependencies['status_aggregator']
 
         status_aggregator.add_provider(provide_status)
 
         api.add_resource(
-            StatusResource, '/status', resource_class_args=[status_aggregator]
+            StatusResource, '/status', resource_class_args=[status_aggregator, config]
```
Comment (fblackburn1) [inline wazo_dird/plugins/status/plugin.py:24]: Do not forget to revert this change when you will know how status aggregator work

File: integration_tests/suite/test_status.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
+# Copyright 2018-2022 The Wazo Authors  (see the AUTHORS file)
```
Comment (fblackburn1) [inline integration_tests/suite/test_status.py:1]: Be careful with copy/paste
```suggestion
# Copyright 2022 The Wazo Authors  (see the AUTHORS file)
```

File: integration_tests/suite/helpers/wait_strategy.py [hunk 8]
```diff
@@ -18,7 +19,12 @@ def is_ready():
                 status = integration_test.dird.status.get()
             except requests.RequestException:
                 status = {}
-            assert_that(status, has_entries({'rest_api': has_entries(status='ok')}))
+            assert_that(
+                status,
+                has_entries(
+                    master_tenant=has_entries(status='ok'),
+                ),
+            )
```
Comment (fblackburn1) [inline integration_tests/suite/helpers/wait_strategy.py:27]: Why removing `rest_api` key? The class name is `RestApiOkWaitStrategy` ...

File: integration_tests/suite/helpers/wait_strategy.py [hunk 9]
```diff
@@ -8,6 +8,7 @@
 from wazo_test_helpers import until
 from wazo_test_helpers.wait_strategy import WaitStrategy, NoWaitStrategy
 
+
```
Comment (fblackburn1) [inline integration_tests/suite/helpers/wait_strategy.py:11]: why?

File: integration_tests/suite/test_auto_create.py [hunk 10]
```diff
@@ -39,7 +39,16 @@ def setUp(self):
...
+            assert_that(
+                response,
+                has_entries(
+                    {
+                        'master_tenant': has_entries(status='ok'),
+                        'rest_api': has_entries(status='ok'),
+                        'bus_consumer': has_entries(status='ok'),
+                    }
+                ),
+            )
```
Comment (fblackburn1) [inline integration_tests/suite/test_auto_create.py:52]: Why waiting for everything?

File: integration_tests/assets/etc/wazo-dird/conf.d/50-default.yml [hunk 11]
```diff
@@ -5,8 +5,7 @@ rest_api:
 
 auth:
   host: auth
-  username: wazo-dird
-  password: secret
+  key_file: /var/lib/wazo-auth-keys/wazo-dird-key.yml
```
Comment (fblackburn1) [inline integration_tests/assets/etc/wazo-dird/conf.d/50-default.yml:8]: Please use username/password
`key_file` is not really cloud friendly

File: integration_tests/assets/docker-compose.yml [hunk 12]
```diff
@@ -17,6 +17,7 @@ services:
     volumes:
       - "../..:/usr/src/wazo-dird"
       - "./etc/wazo-dird/conf.d/50-default.yml:/etc/wazo-dird/conf.d/50-default.yml"
+      - "./keys/wazo-dird-key.yml:/var/lib/wazo-auth-keys/wazo-dird-key.yml:ro"
```
Comment (fblackburn1) [inline integration_tests/assets/docker-compose.yml:20]: remove

File: integration_tests/assets/keys/wazo-dird-key.yml [hunk 13]
```diff
@@ -0,0 +1,2 @@
+service_id: wazo-dird-service
+service_key: whatever-secret-random-key
```
Comment (fblackburn1) [inline integration_tests/assets/keys/wazo-dird-key.yml:2]: remove

File: integration_tests/assets/docker-compose.base.override.yml [hunk 14]
```diff
@@ -0,0 +1,12 @@
...
+  sync:
+    depends_on:
+      - auth
+      - dird
+      - db
+      - rabbitmq
+    environment:
+      TARGETS: "dird:9489 auth:9497 db:5432 rabbitmq:5672"
+      TIMEOUT: "${INTEGRATION_TEST_TIMEOUT}"
+
```
Comment (fblackburn1) [inline integration_tests/assets/docker-compose.base.override.yml:12]: Don't create another asset, we want to have the less possible assets. Use `all_route`

File: integration_tests/suite/test_config.py [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+from .helpers.base import (
+    BaseDirdIntegrationTest,
+    VALID_TOKEN,
+    VALID_SUB_TOKEN,
+    MASTER_TENANT,
+    SUB_TENANT,
+)
+
+
+class TestConfig(BaseDirdIntegrationTest):
```
Comment (fblackburn1) [inline integration_tests/suite/test_config.py:29]: Test are missing when wazo-auth is down
You want to be sure that 503 is triggered

File: integration_tests/suite/test_config.py [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+    assert_that,
+    has_key,
+    calling,
+    has_properties,
+)
+
+
+from wazo_dird_client.exceptions import DirdError
+from wazo_test_helpers.hamcrest.raises import raises
+
```
Comment (fblackburn1) [inline integration_tests/suite/test_config.py:15]:
```suggestion
```

File: integration_tests/suite/test_config.py [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+
+from .helpers.wait_strategy import EverythingOkWaitStrategy
+
```
Comment (fblackburn1) [inline integration_tests/suite/test_config.py:18]: why 2 spaces between import?
```suggestion
```

File: integration_tests/suite/test_config.py [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+
+from hamcrest import (
+    assert_that,
+    has_key,
+    calling,
+    has_properties,
+)
+
```
Comment (fblackburn1) [inline integration_tests/suite/test_config.py:11]:
```suggestion
```

File: integration_tests/suite/helpers/base.py [hunk 19]
```diff
@@ -77,48 +94,78 @@ def restart_postgres(cls):
...
+    service = 'dird'
 
     @classmethod
     def setUpClass(cls):
         super().setUpClass()
-        cls.mock_auth_client = MockAuthClient(
-            '127.0.0.1', cls.service_port(9497, 'auth')
+        cls.host = '127.0.0.1'
+        cls.port = cls.service_port(9489, 'dird')
+        if cls.asset not in ('documentation', 'short-self-stop'):
```
Comment (fblackburn1) [inline integration_tests/suite/helpers/base.py:108]: We should not hardcode asset name here. If you want special configuration for these asset, you just need to create antoher class with BaseDirdIntegration as inheritance

File: integration_tests/suite/helpers/base.py [hunk 20]
```diff
@@ -77,48 +94,78 @@ def restart_postgres(cls):
...
-            auth_port = cls.service_port(9497, 'auth')
-            auth_client = AuthClient('127.0.0.1', auth_port, prefix=None, https=False)
-            auth_client.users.new(
-                uuid=VALID_UUID, tenant_uuid=MAIN_TENANT, username='foobar'
-            )
-        except Exception:
-            pass
-        cls.dird = cls.get_client()
-        cls.wait_strategy.wait(cls)
+    def make_dird(cls, token, tenant=None, **kwargs):
```
Comment (fblackburn1) [inline integration_tests/suite/helpers/base.py:163]: `kwargs` is not unused argument

**Review by fblackburn1 — CHANGES_REQUESTED** [review]
I would like to you answer to each question, not only fixing them
Also can you reword all you `why` commit by following this kind of guideline: https://cbea.ms/git-commit/
The why should be in the description, not the title of the commit

File: wazo_dird/plugins/status/http.py (hunk 5 above)
Comment (nissou91) [reply wazo_dird/plugins/status/http.py:28]: Actually I kept the original at the begining (like wazo-chatd and wazo-setupd); however the `test_config.py` was failing with `master_tenant` key not found. So I took a look at the code in `wazo-webhood`: [status-http](https://github.com/wazo-platform/wazo-webhookd/blob/master/wazo_webhookd/plugins/status/http.py#L15)

File: wazo_dird/plugins/config/http.py (hunk 4 above)
Comment (nissou91) [reply wazo_dird/plugins/config/http.py:13]: did not notice that, I have removed it.

File: wazo_dird/controller.py (hunk 3 above)
Comment (nissou91) [reply wazo_dird/controller.py:41]: Removed it

File: CHANGELOG.md (hunk 1 above)
Comment (nissou91) [reply CHANGELOG.md:11]: It means that `status` plugin takes as a parameter the `config` dict; from which we check if the `master_tenant_uuid` key is set, the status of `master_tenant` is set to `Ok`; otherwise to `fail`: [wazo-webhookd-status](https://github.com/wazo-platform/wazo-webhookd/blob/master/wazo_webhookd/plugins/status/http.py#L12)

File: integration_tests/suite/test_auto_create.py (hunk 10 above)
Comment (nissou91) [reply integration_tests/suite/test_auto_create.py:52]: I have put it back; thought since we are using `EverythingOkWaitStrategy`; we had to wait for all three to have status `Ok`.

File: integration_tests/suite/helpers/wait_strategy.py (hunk 9 above)
Comment (nissou91) [reply integration_tests/suite/helpers/wait_strategy.py:11]: ?

**Comment by sduthil** [comment]
recheck

**Comment by sduthil** [comment]
recheck

**Review by fblackburn1 — CHANGES_REQUESTED** [review]
Please revert your change about renaming MAIN_TENANT to MASTER_TENANT and SUB_TENANT to USERS_TENANT. Then revert token renaming too
Keep main/sub name to avoid any confusion and be explicit on test.
Almost all tests don't care of kind of tenant but only care of the tenant relation between them.
If a test is more explicit with `USERS_TENANT/MASTER_TENANT` then use this kind of syntax
```python
from ..constants import SUB_TENANT as USERS_TENANT
```

**Review by fblackburn1 — APPROVED** [review]
